pkg/registry/blobs.go

```diff
@@ -0,0 +1,110 @@
+package registry
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"strings"
+)
+
+// Returns whether this url should be handled by the blob handler
+// This is complicated because blob is indicated by the trailing path, not the leading path.
+// https://github.com/opencontainers/distribution-spec/blob/master/spec.md#pulling-a-layer
+// https://github.com/opencontainers/distribution-spec/blob/master/spec.md#pushing-a-layer
+func isBlob(req *http.Request) bool {
+	elem := strings.Split(req.URL.Path, "/")
+	elem = elem[1:]
+	if len(elem) < 3 {
+		return false
+	}
+	return elem[len(elem)-2] == "blobs" || (elem[len(elem)-3] == "blobs" &&
+		elem[len(elem)-2] == "uploads")
+}
+
+// blobs
+type blobs struct {
+	// Blobs are content addresses. we store them globally underneath their sha and make no distinctions per image.
+	contents map[string][]byte
+	// Each upload gets a unique id that writes occur to until finalized.
+	uploads map[string][]byte
+}
+
+// https://github.com/opencontainers/distribution-spec/blob/master/spec.md#pulling-a-layer
+// https://github.com/opencontainers/distribution-spec/blob/master/spec.md#pushing-a-layer
+func (b *blobs) handle(resp http.ResponseWriter, req *http.Request) {
+	elem := strings.Split(req.URL.Path, "/")
+	elem = elem[1:]
+	// Must have a path of form /v2/{container}/blobs/{upload,sha256:}
+	if len(elem) < 4 {
+		resp.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	target := elem[len(elem)-1]
+	service := elem[len(elem)-2]
```

Comment:
Why `service`?

Reply:
I needed a name and calling it upload service seemed to work? If you have a suggestion I'm all ears.